### DISCUSSION TITLE
objects.Hist: how to get a log scale for counts?

### DISCUSSION THREAD

**Discussion by tburnett** [mwaskom/seaborn#3401] [Q&A]
I'm trying to duplicate the sns.histplot log_scale  option to scale the counts.  Or log=True in plt.hist. 
The histogram disappears after a  Plot.scale(y='log') . 
Before and after:
![image](https://github.com/mwaskom/seaborn/assets/1407165/82b857fd-5c5a-46bb-9511-1be0cdd6973d)
![image](https://github.com/mwaskom/seaborn/assets/1407165/3d749887-686e-470c-bf77-67b4a3a9637a)

**Comment by thuiop** [comment]
Yes, this is a known issue ; has to do with how the coordinates system works (notice there is no "0" in the log scale). See #2454 for more details. In the meantime, you should avoid using `Plot.scale` for this specific case, and use `ax.set_yscale("log")` manually.

**Reply by tburnett** [reply]
Background: I love the elegance of this object model, and its implementation, and am exploring the feasibility of switching from my matplotlib-only practice. The documentation has lots of nice examples focusing on what one can do with a single dataframe, but little beyond that. This issue and the inflexibility of the legend are potential show-stoppers unless I find not-too-klugy workarounds. Access to the `plt.Figure` object before rendering, as you suggest, is essential. This would be in the expectation that there is ongoing development to eventually address the issues.

The implementation creates the figure with the plot() method, 
Here is what I came up with:
```
import seaborn.objects as so
import numpy as np
from scipy.stats import norm
r = np.random.normal(size=10000)
x = np.linspace(-4,4,100)
y = norm.pdf(x)

ptr =(so.Plot()
     .layout(size=(4,2)).limit(x=(-4,4))
     .add(so.Bars(), so.Hist(stat='density', bins=25), x=r)
     .add(so.Line(color='red'), x=x, y=norm.pdf(x))
     .plot()
)
ptr._figure.axes[0].set_yscale('log')
ptr
```
![image](https://github.com/mwaskom/seaborn/assets/1407165/fbff1df6-2171-4777-b18e-13dad2a69f28)
The kluge is the need to access the "private" property `_figure`.  It would be less so if `plot` accepted a function to modify the resulting `Figure`. e.g., `plot(lambda fig: fig.axes[0].set(yscale='log')`.
It would be nice to add a legend to this example--this is a case where matplotlib is very convenient, collecting "label" keywords. 
And one other observation: When IPython renders the output of a code cell, it detects the presence of a `Figure`, and renders it in the output. Here it is necessary to put a reference to the `Plotter` object  at the end of the cell, where its `_repr_png_` method gets automatically invoked.

**Reply by thuiop** [reply]
You can provide it a figure or an ax by using `Plot.on(fig)` or `Plot.on(ax)` before using `Plot.plot()`. For the legend, this is definitely a pain currently, partly because matplotlib is also lacking a bit in this respect.

**Reply by tburnett** [reply]
Thanks for reminding me about `Plot.on`, an option I should have considered. 
However, my desire is to be able to make a simple change in the specification of `so.Plot` itself to turn on the y-axis log scale.
Playing with a `so.Plot` subclass (for which I had to add its own `_clone()`), with an overridden `plot`, I can do it as I suggested, modifying the figure created by `Plot.plot`.

About the legend, is there no work-around/kluge? If not, I'm afraid I'll have to drop this for now and wait.

**Reply by thuiop** [reply]
Depends on what you want to do exactly with the legend.

**Reply by tburnett** [reply]
Legend: In matplotlib, one has a "label=..." for each line or hist, and a fig.legend()  collects them all.
```
import matplotlib.pyplot as plt
fig, ax = plt.subplots(figsize=(4,2))
ax.hist(r, np.linspace(-3,3,25), density=True, label='random.normal')
ax.plot(x,y, label='norm(0,1)')
fig.legend()
```
![image](https://github.com/mwaskom/seaborn/assets/1407165/96e4da8d-7fa7-493a-a258-04e2789378e1)

**Reply by thuiop** [reply]
Ah, I see. This is probably pretty complex to do ; I would probably just add a separate legend in that case.

**Reply by tburnett** [reply]
Not sure what you mean by "separate".  Again, if I had a private way to produce a MPL-like legend, I'd continue exploring how to apply this system to my every-day efforts. I've already understood how to make a Plot subclass, which I use to get a log scale.  (Minor, but It would be nice it `Plot.clone` could return an instance of its actual class.) 
Related question: is there an example somewhere showing how to duplicate the MPL errorbar display?

**Reply by thuiop** [reply]
The legend seaborn creates is a matplotlib (Figure) legend. The only issue is that it cannot be modified afterwards without doing some weird stuff. However, matplotlib does support having several legends, so you can make a second "box" with the line you drew afterwards for instance.

**Comment by mwaskom** [comment]
Dupe of https://github.com/mwaskom/seaborn/issues/3032?

**Comment by tburnett** [comment]
It doesn't look like it. Adding a "a. scale(y='symlog')" does not work.

This is a very common use case in physics and astrophysics, overlaying a distribution with a function or functions. Publication-quality plots demand legends.

--Toby Burnett
________________________________

**Reply by mwaskom** [reply]
Your post does not include a reproducible example so I really have no clue what your actual issue is. Telling me that it’s important in physics is pointless if you can’t actually articulate the problem.

**Reply by tburnett** [reply]
OK, 
```
import seaborn.objects as so
import numpy as np
from scipy.stats import norm

r = np.random.normal(size=100000)
x = np.linspace(-4,4,100)
y = norm.pdf(x)

p =(so.Plot()
     .layout(size=(4,2)).limit(x=(-4,4))
     .add(so.Bars(), so.Hist(stat='density', bins=25), x=r)
     .add(so.Line(color='red'), x=x, y=y)
)

p.show(), p.scale(y='log').show()
```
